### PR TITLE
Skip Updating CopyComplete Marker When Not Necessary

### PR DESCRIPTION
**Pull request by BenVillalobos**
Fixes #https://github.com/dotnet/msbuild/issues/6576

### Context
https://github.com/dotnet/msbuild/issues/6576 revealed that the `.copycomplete` file marker is updated even when the `Copy` task in `_GetCopyFilesMarkedLocal` doesn't _actually_ copy anything. This can mess with incremental builds.

### Changes Made
This change adds an output parameter, `CopiedAtLeastOneFile` to the `Copy` task that the `Touch` task is now conditioned off of.

### Testing
Tested local builds

### Notes
This could also be done by having an ITaskItem[] that contains all files that were actually copied. Then the touch task could check if that item were empty. I opted for the straightforward route since the ITaskItem[] solution isn't needed yet, and this implementation can easily be changed when we do need that.

### CODE REVIEW

**Comment by BenVillalobos** [comment]
Much like rainer suggested [here](https://github.com/dotnet/msbuild/issues/6576#issuecomment-872340678), I don't think this fixes some other incremental build problems in CSWinRT, but this was certainly a bug for CopyComplete in the first place.

cc @scottj1s

**Review by Forgind — CHANGES_REQUESTED** [review]
I don't think this is right. First, I imagine things that don't technically move anything around but create links still should have it marked as if something were copied. More importantly, the copy step itself sometimes throws an exception if the file doesn't need to be copied, so CopiedAtLeastOneFile should be set after that.

File: src/Tasks/Copy.cs [hunk 1]
```diff
@@ -297,6 +300,12 @@ private void LogDiagnostic(string message, params object[] messageArgs)
                 Log.LogMessage(MessageImportance.Normal, FileComment, sourceFilePath, destinationFilePath);
 
                 File.Copy(sourceFileState.Name, destinationFileState.Name, true);
+                CopiedAtLeastOneFile = true;
+            }
+            else
+            {
+                // Properly linked files will be considered successfully copied files.
+                CopiedAtLeastOneFile = true;
             }
```
Comment (Forgind) [inline src/Tasks/Copy.cs:309]:
```suggestion
            }

            // Files were successfully copied or linked. Those are equivalent here.
           CopiedAtLeastOneFile = true;
```

? Little shorter.

**Review by Forgind — APPROVED** [review]
Looks good to me!

File: src/Tasks/Copy.cs [hunk 2]
```diff
@@ -299,6 +302,9 @@ private void LogDiagnostic(string message, params object[] messageArgs)
                 File.Copy(sourceFileState.Name, destinationFileState.Name, true);
             }
 
+            // Files were successfully copied or linked. Those are equivalent here.
+           CopiedAtLeastOneFile = true;
```
Comment (ladipro) [inline src/Tasks/Copy.cs:306]: super-nit: This line is not indented properly.

**Review by ladipro — COMMENTED** [review]
> This could also be done by having an ITaskItem[] that contains all files that were actually copied.

Why doesn't `CopiedFiles` work here?

https://github.com/dotnet/msbuild/blob/65e44d936dcfbac421d0e786dac68ee22edb1d86/src/Tasks/Copy.cs#L136-L140

**Comment by ladipro** [comment]
> Why doesn't CopiedFiles work here?

Found an answer in the issue! Wondering if the name of the new parameter shouldn't be more explicit in encoding the difference between it and non-empty `CopiedFiles`.

**Comment by BenVillalobos** [comment]
@ladipro Any suggestions? I considered `ActuallyCopiedAtLeastOneFile` but thought that may have been too snarky 😄

**Comment by ladipro** [comment]
Maybe `SuccessfullyCopiedAtLeastOneFile`?

What **exactly** is the difference between non-empty `CopiedFiles` and this new output parameter being true? That should help come up with the best name, I believe.

**Comment by Forgind** [comment]
CopiedFiles is nonempty if either it copies files (in which case this is true) or it notices it doesn't have to copy files because it's up-to-date, in which case this would be false.

**Comment by ladipro** [comment]
Thank you, how about `UpdatedAtLeastOneFile` then? Plus a good comment on the prop.

**Review by ladipro — APPROVED** [review]
Thank you!

**Comment by Forgind** [comment]
ChangedFileContents? UpdatedAtLeastOneFile feels to me like it's specifically overwriting, whereas this could be making a new file where there wasn't one before.

**Comment by ladipro** [comment]
`WroteAtLeastOneFile`? `ChangedFileContents` sounds like a prop to hold "contents", i.e. it's not obvious that it's a boolean. But not feeling strongly about it.

**Comment by Forgind** [comment]
It isn't a huge deal—better would've been to change CopiedFiles, but that isn't an option. I know I've also been annoyed sometimes when there was too much haggling over names 🙂

Whatever you decide, I'm fine with @BenVillalobos

**Comment by BenVillalobos** [comment]
> I know I've also been annoyed sometimes when there was too much haggling over names

Yeah it gets particularly haggle-some when its public and therefore shouldn't be changed once it's in. `UpdatedAtLeastOneFile` makes slightly less sense when creating a new file. `WroteAtLeastOneFile` makes sense for both creating a link and creating a new file. Going with `WroteAtLeastOneFile` as the final name.

**Pull request merged**